genesis_core/node/scheduler/service.py

```diff
@@ -169,6 +185,92 @@ def _schedule_nodes(self) -> tp.List[models.Machine]:
+        ]
+        vms = []
+
+        for unscheduled_node in unscheduled:
+            node = unscheduled_node.node
+            if node.node_type == nc.NodeType.HW:
+                idle_machines = idle_hws
+            else:
+                idle_machines = idle_vms
+
+            # Filtering. We filter out unsuitable machines. For instance,
+            # machines that doesn't have enough cores or ram or some
+            # placement constraints.
+            for filter in self._machine_filters:
+                idle_machines = filter.filter(node, idle_machines)
+
+            # There are no available HW machines for this node
+            # This means we unable to proceed scheduling process
+            # for this node.
+            if not idle_machines and node.node_type == nc.NodeType.HW:
+                LOG.warning(
+                    "No HW machines found to schedule node %s",
+                    node.uuid,
+                )
+                if node.status != nc.NodeStatus.ERROR:
+                    node.status = nc.NodeStatus.ERROR.value
+                    node.description = "No suitable HW machines found"
+                    node.save()
+                continue
+
+            # There are no available VM machines for this node
+            # but it's not a problem. A virtual machine will be
+            # created later.
+            if not idle_machines and node.node_type == nc.NodeType.VM:
+                LOG.debug(
+                    "No idle VM machines found to schedule node %s",
+                    node.uuid,
+                )
+                vms.append(node)
+                continue
+
+            # Weighting. We weight machines and choose the best one.
+            # Accumulate weights from all weighters.
+            # So that the best pool has the highest weight
+            accumulated_weights = [0.0] * len(idle_machines)
+            for weighter in self._machine_weighters:
+                weights = weighter.weight(idle_machines)
+                accumulated_weights = [
+                    w0 + w1 for w0, w1 in zip(accumulated_weights, weights)
+                ]
+
+            # Choose the best machine, it means the one with the highest weight
+            index = accumulated_weights.index(max(accumulated_weights))
+            machine = idle_machines[index]
+
+            machine.node = node.uuid
+            machine.status = nc.MachineStatus.SCHEDULED.value
+            node.status = nc.NodeStatus.SCHEDULED.value
+            try:
+                node.save()
+                machine.save()
+                LOG.info(
+                    "The node %s scheduled to %s machine",
+                    node.uuid,
+                    machine.uuid,
+                )
+            except Exception:
+                LOG.exception("Error scheduling node %s", node.uuid)
+
+            # Actualize idle machines
+            if node.node_type == nc.NodeType.HW.value:
+                idle_hws.remove(machine)
```

Comment:
idle_hws/idle_vms is not returned to anywhere, and it's used only as `idle_machines = idle_hws`. Maybe I miss something, but looks like we can just drop these removals, no?